File: BaekJoon/Gold/10830/Python/10830.py
# no. 10830: 행렬 제곱 (Gold IV)
# 분할정복
# 입력으로 주어지는 행렬의 각 숫자가 1000보다 "작거나 같은 것에" 주의하자.

from sys import stdin
input = stdin.readline
r = 1000

n, b = map(int, input().split())
matrix = [[num%r for num in map(int, input().split())] for _ in range(n)]

def multiple(a, b):
    new_matrix = [[0 for _ in range(n)] for _ in range(n)]
    for i in range(n):
        for j in range(n):
            for k in range(n):
                new_matrix[i][j] += (a[i][k]%r)*(b[k][j]%r)%r  
            new_matrix[i][j] %= r
    return new_matrix

def power(a, n):
    if n == 1:
        return a
    tmp = power(a, n//2)
    if n % 2:
        return multiple(multiple(tmp, tmp), a)
    else:
        return multiple(tmp, tmp)

result = power(matrix, b)
for arr in result:
    print(*arr)

File: BaekJoon/Gold/10830/C++/10830.cpp
// no.10830: 행렬 제곱 (G4)

#include <cstdio>
#include <vector>
#define R 1000
using namespace std;

typedef long long int64;

int n;
vector<vector<int>> matrix;
vector<vector<int>> power(int64);
vector<vector<int>> multiply(vector<vector<int>>, vector<vector<int>>);

int main() {
    int64 b;
    scanf("%d %lld", &n, &b);
    matrix.assign(n, vector<int>(n, 0));
    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            int num;
            scanf("%d", &num);
            matrix[i][j] = num%R;
        }
    }
    vector<vector<int>> ret = power(b);
    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            printf("%d ", ret[i][j]);
        }
        printf("\n");
    }
    return 0;
}

vector<vector<int>> power(int64 b) {
    if(b == 1) {
        return matrix;
    }
    vector<vector<int>> tmp = power(b/2);
    if(b%2 == 0)
        return multiply(tmp, tmp);
    else
        return multiply(multiply(tmp, tmp), matrix);
}

vector<vector<int>> multiply(vector<vector<int>> a, vector<vector<int>> b) {
    vector<vector<int>> ret(n, vector<int>(n, 0));
    for(int i=0; i<n; i++)
        for(int j=0; j<n; j++)
            for(int k=0; k<n; k++)
                ret[i][j] = (ret[i][j] + a[i][k]*b[k][j]) % R;
    return ret;
}